### DISCUSSION TITLE
Modifying objects on branch from Netbox custom scripts

### DISCUSSION THREAD

**Discussion by qspark-eliezerlp** [netboxlabs/netbox-branching#51] [Q&A] [answered]
Hi all 👋 

Since https://github.com/netboxlabs/netbox-branching/issues/42 was recently merged, I loaded up the develop branch (@[commit](https://github.com/netboxlabs/netbox-branching/commit/61bc9420a3df59d9384618e543fe0cb280f05b6e)) and was able to load and run custom scripts.

Here is an example of the kinds of interactions being done within the scripts:
```
interface, created = Interface.objects.get_or_create(device=parent.device, name=interface_name)

if created:
  self.log_info(f"Creating new interface {interface_name} for {parent.device}")
else:
  self.log_info(f"Updating existing interface {interface_name} for {parent.device}")

# Snapshot to generate the correct change log data when editing an existing object
interface.snapshot()

interface.type = InterfaceTypeChoices.TYPE_VIRTUAL
interface.save()

# Ensure the interface fields are correct
interface.mac_address = parent.mac_address
interface.enabled     = True
interface.mode        = InterfaceModeChoices.MODE_TAGGED
interface.parent      = parent

interface.full_clean()
interface.save()

self.log_success(f"Interface changes saved: {interface}")
```

The script is able to run successfully and modifies the objects on the Main branch. Is there a way to direct the script to modify the objects on the currently active branch?

I did notice the following mentions in a recent [blog post](https://netboxlabs.com/blog/netbox-branching-public-beta/) which _might_ relate:
> - _Once the interfaces for NetBox Branching have settled we’ll start looking at adding support for our integrations and SDKs_
> - _If you’re a NetBox plugin maintainer, we’ll be running invite-only informational sessions over the next couple of weeks with insights into how you could take advantage of Branching in your plugins. Keep an eye on your inbox for an email about those._

**Comment by qspark-eliezerlp** [comment, marked as answer]
In the recent blog post **_[NetBox Branching Enters General Availability](https://netboxlabs.com/blog/netbox-branching-enters-general-availability/)_** the following is mentioned, seemingly confirming this is currently unsupported:

> - Formal API support for use in custom scripts

**Comment by RedLine89** [comment]
Really looking forward to this API support! I assume there is no clarity on when potentially this could be available?

**Reply by qspark-eliezerlp** [reply]
I believe it will be covered by #9. I am not clear if the API already exists but is simply not documented yet.

**Reply by RedLine89** [reply]
Thank you for confirming.

Yeah, I mocked around a bit with it and it seems possible to access branch data from a custom script. So perhaps it is indeed just documentation that is left.

```
from netbox_branching.models.branches import Branch

        branch = Branch.objects.get(name="port_update")
        unmerged_changes = branch.get_unmerged_changes().order_by('time')
        for item in unmerged_changes:
            self.log_debug(f"Changed object: {item.changed_object.}")
```

**Comment by jasonjuenger** [comment]
It isn't documented yet, but you can accomplish this by accessing the schema_id in the `active_branch` cookie from the request, getting the branch object from the schema id, then using the `activate_branch` context to make your changes.

```
branch_schema_id = self.request.COOKIES.get('active_branch')
branch = Branch.objects.get(schema_id=branch_schema_id)
with activate_branch(branch):
    # make changes here
```
